store/tikv/store_aliveness.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pingcap/failpoint"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/pdapi"
+	"go.uber.org/zap"
+	"golang.org/x/sync/singleflight"
+)
+
+const knobSwitch = "livenessKnob"
+
+// StoreLiveness maintains the liveness of store.
+type StoreLiveness struct {
+	sf             singleflight.Group
+	timeout        time.Duration
+	statusResolver *kvStatusAddrResolver
+	doneCh         chan struct{}
+	knobs          struct {
+		mockTiKVStatus func(statusAddr string) error
+	}
+}
+
+func newStoreLiveness(pdHosts []string, timeout time.Duration) *StoreLiveness {
+	if len(pdHosts) == 0 {
+		return nil
+	}
+	doneCh := make(chan struct{})
+	l := &StoreLiveness{
+		timeout:        timeout,
+		statusResolver: newStatusAddressResolver(pdHosts, doneCh),
+		doneCh:         doneCh,
+	}
+	return l
+}
+
+// IsAlive checks whether store process still alive.
+func (s *StoreLiveness) IsAlive(storeID uint64) bool {
+	if s == nil {
+		return false
+	}
+	_, err, _ := s.sf.Do(strconv.FormatUint(storeID, 10), func() (interface{}, error) {
+		err := s.checkKvStatusAPI(storeID)
+		if err != nil {
+			logutil.BgLogger().Warn("[liveness] send fail and store is down, need invalid region cache and switch next peer",
+				zap.Uint64("store", storeID), zap.Error(err))
+			return nil, err
+		}
+		return nil, nil
+	})
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+func (s *StoreLiveness) start() error {
+	failpoint.Inject(knobSwitch, func() {
+		if s.statusResolver.knobs.mockStart != nil {
+			failpoint.Return(s.statusResolver.knobs.mockStart(s.statusResolver))
+		}
+	})
+	return s.statusResolver.start()
+}
+
+func (s *StoreLiveness) close() {
+	if s == nil {
+		return
+	}
+	close(s.doneCh)
+}
+
+func (s *StoreLiveness) checkKvStatusAPI(storeID uint64) error {
+	statusAddr := s.statusResolver.resolve(storeID)
+	if len(statusAddr) == 0 {
+		return errors.Errorf("unknown store %d maybe caused by offline in pd or cache outdated", storeID)
+	}
+	failpoint.Inject(knobSwitch, func() {
+		if s.knobs.mockTiKVStatus != nil {
+			failpoint.Return(s.knobs.mockTiKVStatus(statusAddr))
+		}
+	})
+	ctx, cancel := context.WithTimeout(context.Background(), s.timeout)
```

Comment:
50ms is too small, I think 120ms may be a better default value.

Reply:
We have to take 2 network roundtrips and system scheduler, and GC and all other things into consideration.